### ISSUE TITLE
Identify validations to perform for every action

### ISSUE THREAD

**Issue by alelb** [fabbricadigitale/scimd#7]
Add a table to [Docs/Design.md](https://github.com/fabbricadigitale/scimd/blob/develop/docs/DESIGN.md) to show validations vs actions

**Comment by alelb** [comment]
The following table shows validations to perform per action.

|   |  create | read  | replace  | delete  | update |
|---|---|---|---|---|---|
| **type**  | :heavy_check_mark:  | | :heavy_check_mark:  | | :heavy_check_mark: |
| **required**  | :heavy_check_mark:  | | :heavy_check_mark:  | | :heavy_check_mark:  | 
| **uniqueness**  | :heavy_check_mark:  |  | :heavy_check_mark:  |  | :heavy_check_mark:  |
| **mutability**  | *  | **  | :heavy_check_mark:  |  | :heavy_check_mark:  |


### Type validation
Checks whether the type of the attributes is consistent with the one defined in the schema. Performed by **Data Model** and **Unmarshal** operation.

### Required validation
Checks whether the mandatory attributes have been provided. [RFC7644](https://tools.ietf.org/html/rfc7644#section-3.12)

### Uniqueness
Avoid storing a resource that has a unique attribute value already in use.
[RFC7644](https://tools.ietf.org/html/rfc7644#section-3.12)

### Mutability
Checks whether the attempted modification is not compatible with the target attibute's mutability.

(*) In a **Create** resource operation Attributes whose mutability is "readWrite" and that are omitted from the request body MAY be assumed to be not asserted by the client.  The service provider MAY assign a default value to non-asserted attributes in the final resource representation. [RFC7644](https://tools.ietf.org/html/rfc7644#section-3.3)

(**) In a **Read** resource operation Attributes whose mutability is "writeOnly" are omitted from the server response.

In a **Update** or a **Replace** operations if the request breaks mutability rule the server response with a status code 400 (Bad Request). [RFC7644](https://tools.ietf.org/html/rfc7644#section-3.12)

### Attributes syntax
*TBD*

**Comment by leogr** [comment]
Type validation when **Unmarshalling** has been implemented by [decode.go](https://github.com/fabbricadigitale/scimd/blob/develop/schemas/core/decode.go) for *mapped resource*, instead *structured* ones are using json tags.

**Comment by leogr** [comment]
For read ops, "writeOnly" mutability characteristic should be accomplished already by [attr.Projection api used here](https://github.com/fabbricadigitale/scimd/blob/develop/api/query/resource.go#L68).
@alelb Could you check that the API above is working properly for this case?

**Comment by alelb** [comment]
[rfc7643](https://tools.ietf.org/html/rfc7643#section-7) said that `An attribute with a mutability of "writeOnly" usually also has a returned setting of "never"`. 
@leogr  Despite the "usually" adverb, we can add a validator to verify that.

**Comment by leogr** [comment]
Damn RFC 👎  
@alelb I prefer to do not implement a validator for that.

Also, RFC says:
```
 writeOnly  The attribute MAY be updated at any time.  Attribute
            values SHALL NOT be returned (e.g., because the value is a
            stored hash).  Note: An attribute with a mutability of
            "writeOnly" usually also has a returned setting of "never".
```
So, attribute values **MUST NOT** be returned (it's mandatory).
Thus, for security purposes (eg. the password) too, it may make sense to do not fetch those attributes ("writeOnly" ones) from the db, but implications must be evaluated.
Furthermore, we may implement both "mutability" and "uniqueness" into the storage layer (BTW, I can't figure any other place where "uniqueness" could be enforced).

**Comment by leogr** [comment]
However, mutability "writeOnly" can be also implemented easily within the [projection API](https://github.com/fabbricadigitale/scimd/blob/develop/api/attr/projection.go#L9) just adding those attributes to the never-returned set.

**Comment by leodido** [comment]
What is the state of this topic guys?

cc @leogr @alelb

**Comment by leogr** [comment]
@leodido 
`type` and `required` validation are built-in resource data handling.
`mutability` for `read` ops was fixed by 95ffe0d
 Ops other than `read` are not yet implemented.
However the approach is simply:
- `type` / `required` are handled with resource data (marshalling, enforcement, datatype etc..)
- `mutability` must be handled in each API (ie. within the `api` package)
- `uniqueness` could be handled within the `storage` layer only

Mind that those above are not canonical validation tasks, but actually business logic required by RFC.
However, don't forget that formal validation for `api.messages` structs must be added too within API by using `validation.Validator`
